### PR TITLE
(kue) Enable the use of the three-parameter ProcessCallback

### PR DESCRIPTION
**Pull request by googol**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/Automattic/kue#pause-processing https://github.com/Automattic/kue/blob/master/lib/queue/worker.js#L229
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/kue/index.d.ts [hunk 1]
```diff
@@ -33,7 +33,8 @@ export declare class Queue extends events.EventEmitter {
     checkActiveJobTtl(ttlOptions: Object): void;
     watchStuckJobs(ms: number): void;
     setting(name: string, fn: Function): Queue;
-    process(type: string, n?: number | ProcessCallback, fn?: ProcessCallback): void;
+    process(type: string, fn?: ProcessCallback): void;
+    process(type: string, n: number, fn?: ProcessCallback): void;
```
Comment (SimonSchick) [inline types/kue/index.d.ts]: Just `n`, is there anything more descriptive?

Comment (googol) [reply types/kue/index.d.ts]: Well the source code uses `n` as the parameter name. It refers to the number of concurrent workers though.

What do you propose?

Comment (deerawan) [inline types/kue/index.d.ts]: should be fine I think using `n`. 👍

File: types/kue/index.d.ts [hunk 2]
```diff
@@ -63,7 +64,9 @@ interface Priorities {
 
 export type DoneCallback = (err?: any, result?: any) => void;
 export type JobCallback = (err?: any, job?: Job) => void;
-export type ProcessCallback = (job: Job, cb: DoneCallback) => void;
+export type ProcessCallback =
+    | ((job: Job, cb: DoneCallback) => void)
+    | ((job: Job, ctx: WorkerCtx, cb: DoneCallback) => void);
```
Comment (deerawan) [inline types/kue/index.d.ts]: should we need to add a test regarding the changes?

File: types/kue/index.d.ts [hunk 3]
```diff
@@ -144,6 +147,13 @@ declare class Worker extends events.EventEmitter {
     resume(): boolean;
 }
 
+interface WorkerCtx {
+    pause(fn?: DoneCallback): void;
+    pause(timeout: number, fn?: DoneCallback): void;
+    resume(): void;
+    shutdown(): void;
+}
```
Comment (deerawan) [inline types/kue/index.d.ts]: looks good!

File: types/kue/index.d.ts (hunk 2 above)
Comment (googol) [reply types/kue/index.d.ts]: Yes, that's a good idea. I added the pausing example from the readme as a test

**Comment by googol** [comment]
@deerawan Added the test

**Review by deerawan — APPROVED** [review]

**Comment by googol** [comment]
Thanks!

**Pull request merged**